### PR TITLE
Enable test suite for windows, ghc-8.8 and ghc-8.10

### PR DESCRIPTION
**Pull request by jneira**
* `ghcide restarts shake session on config changes` is fixed
* some more tests ignored in ghc-8.8 (related with th)
* succesfull builds (13 consecutives :smile:) :
  * from https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=905
  * to https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=917 (still in progress)
* all build takes around 30 minutes, most a little less

### CODE REVIEW

File: .azure/linux-stack.yml [hunk 1]
```diff
@@ -47,6 +47,6 @@ jobs:
     displayName: 'stack build --test --only-dependencies'
   - bash: |
       export PATH=/opt/cabal/bin:$PATH
-      stack test --ghc-options=-Werror  --stack-yaml=$STACK_YAML || stack test --ghc-options=-Werror --stack-yaml=$STACK_YAML --ta "--rerun" || LSP_TEST_LOG_COLOR=0 LSP_TEST_LOG_MESSAGES=true LSP_TEST_LOG_STDERR=true stack test --ghc-options=-Werror --stack-yaml=$STACK_YAML --ta "--rerun"
+      stack test --ghc-options=-Werror  --stack-yaml=$STACK_YAML --ta "--rerun" || stack test --ghc-options=-Werror --stack-yaml=$STACK_YAML --ta "--rerun" || LSP_TEST_LOG_COLOR=0 LSP_TEST_LOG_MESSAGES=true LSP_TEST_LOG_STDERR=true stack test --ghc-options=-Werror --stack-yaml=$STACK_YAML --ta "--rerun"
```
Comment (ndmitchell) [inline .azure/linux-stack.yml:50]: This change for consistency? What does rerun the first time around mean?

File: .azure/windows-stack.yml [hunk 2]
```diff
@@ -5,10 +5,10 @@ jobs:
     vmImage: 'windows-2019'
   strategy:
     matrix:
-#     stack_810:
-#       STACK_YAML: "stack8101.yaml"
-#     stack_88:
-#       STACK_YAML: "stack88.yaml"
+      stack_810:
+       STACK_YAML: "stack8101.yaml"
```
Comment (ndmitchell) [inline .azure/windows-stack.yml:9]: The name 8101 is very hard t parse. Could we have stack_8_10_1.yaml?

File: test/exe/Main.hs [hunk 3]
```diff
@@ -3477,6 +3485,7 @@ clientSettingsTest = testGroup "client settings handling"
             logNot <- skipManyTill anyMessage loggingNotification
             isMessagePresent "Updating Not supported" [getLogMessage logNot]
     ,   testSession "ghcide restarts shake session on config changes" $ do
+            void $ skipManyTill anyMessage $ message @RegisterCapabilityRequest
```
Comment (ndmitchell) [inline test/exe/Main.hs:3488]: Why is this required?

**Review by ndmitchell — COMMENTED** [review]
Looks awesome. A few questions I'd like to understand first, but really nice work.

File: .azure/linux-stack.yml (hunk 1 above)
Comment (jneira) [reply .azure/linux-stack.yml:50]: `tasty` only generates the needed log file if you set `--rerun` (or `--rerun-update`), and it can only filter tests in a second execution with `--rerun` if that log exists (tested locally in windows). 
It will execute the entire test suite if that log has no errors or does not exist. As the log is not saved between ci builds and that is the way you make the second test execution to handle only failing tests. As all calls are chained with `||`, if the first execution is succesful, no more executions will be done.

File: .azure/windows-stack.yml (hunk 2 above)
Comment (jneira) [reply .azure/windows-stack.yml:9]: Or `stack-8.10.1.yaml` like hls. But maybe we should change all stack.yaml's for consistency and if it is the case i would prefer to do it in another pr.

File: .azure/linux-stack.yml (hunk 1 above)
Comment (jneira) [reply .azure/linux-stack.yml:50]: Maybe use each concrete flag (`--rerun-update` in the first one, `--rerun-filter` in second and third) in each execution would make it clearer

Comment (ndmitchell) [reply .azure/linux-stack.yml:50]: I'd much rather the first flag to be `--rerun-create-afresh` or similar - we do caching on the CI, so I worry about stale test success being cached. While it probably isn't, using different flags would make it clear.

File: .azure/windows-stack.yml (hunk 2 above)
Comment (ndmitchell) [reply .azure/windows-stack.yml:9]: Agreed - not really something for this PR

File: .azure/linux-stack.yml (hunk 1 above)
Comment (jneira) [reply .azure/linux-stack.yml:50]: `--rerun-update` creates or replaces the log so i think it would be like `--rerun-create-afresh`

File: test/exe/Main.hs (hunk 3 above)
Comment (jneira) [reply test/exe/Main.hs:3488]: Good question, i added it from pure empirical observation of existing failing tests. But i just reproduced locally the error with this version so it is not valid.
So i have to investigate and replace it.

Comment (pepeiborra) [reply test/exe/Main.hs:3488]: maybe change the test to load a file and wait for diagnostics before sending the WorkspaceDidChangeConfiguration, just to make sure that ghcide has started a build session (otherwise there will be no build session to restart)

Comment (jneira) [reply test/exe/Main.hs:3488]: The session is initialized and restarted, i thought it needed a file too but then i removed it and the session logging notifications still were there (not sure why).

The log of a failing one (before waiting for ReqCaps):

<details><summary>Failing log</summary>

```
 <-- {
     "tag": "RspInitialize",
     "contents": {
.............................
     }
 }
 --> {
     "jsonrpc": "2.0",
     "params": {},
     "method": "initialized"
 }
 --> {
     "jsonrpc": "2.0",
     "params": {
         "settings": ""
     },
     "method": "workspace/didChangeConfiguration"
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": { 
             "type": 1,
             "message": "haskell-lsp:configuration parse error. NotificationMessage {_jsonrpc = \"2.0\", _method = WorkspaceDidChangeConfiguration, _params = DidChangeConfigurationParams {_settings = String \"\"}} \"Updating Not supported\""
         },
         "method": "window/logMessage"
     }
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "type": 1,
             "message": "haskell-lsp:configuration parse error. NotificationMessage {_jsonrpc = \"2.0\", _method = WorkspaceDidChangeConfiguration, _params = DidChangeConfigurationParams {_settings = String \"\"}} \"Updating Not supported\""
         },
         "method": "window/logMessage"
     }
 }
 <-- {
     "tag": "ReqRegisterCapability",
     "contents": {
 -----------------------------------
     }
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "type": 4,
             "message": "Restarting build session (aborting the previous one took 0.00s)"
         },
         "method": "window/logMessage"
     }
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "type": 4,
             "message": "Restarting build session (aborting the previous one took 0.00s)"
         },
         "method": "window/logMessage"
     }
 }
 <-- {
     "tag": "ReqWorkDoneProgressCreate",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "token": "1"
         },
         "method": "window/workDoneProgress/create",
         "id": 1
     }
 }
 <-- {
     "tag": "NotWorkDoneProgressBegin",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "value": {
                 "kind": "begin",
                 "title": "Processing"
             },
             "token": "1"
         },
         "method": "$/progress"
     }
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "type": 4,
             "message": "Finishing build session(exception: AsyncCancelled)"
         },
         "method": "window/logMessage"
     }
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "type": 4,
             "message": "Finishing build session(exception: AsyncCancelled)"
         },
         "method": "window/logMessage"
     }
 }
 <-- {
     "tag": "NotWorkDoneProgressEnd",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "value": {
                 "kind": "end"
             },
             "token": "1"
         },
         "method": "$/progress"
     }
 }
 <-- {
     "tag": "NotLogMessage",
     "contents": {
         "jsonrpc": "2.0",
         "params": {
             "type": 4,
             "message": "finish: kick (took 0.00s)"
         },
         "method": "window/logMessage"
     }
 }
 --> {
     "jsonrpc": "2.0",
     "params": null,
     "method": "shutdown",
     "id": 0
 }
 <-- {
     "tag": "RspShutdown",
     "contents": {
         "result": null,
         "jsonrpc": "2.0",
         "id": 0
     }
 }
 --> {
     "jsonrpc": "2.0",
     "params": [],
     "method": "exit"
 }
 FAIL
       Exception: Timed out waiting to receive a message from the server.
       Last message received:
       {
           "tag": "NotLogMessage",
           "contents": {
               "jsonrpc": "2.0",
               "params": {
                   "type": 4,
                   "message": "finish: kick (took 0.00s)"
               },
               "method": "window/logMessage"
           }
       }
```
</details>

There are one two logging notifications about `Updating Not supported` but only one `didChangeConfiguration` request. :thinking:

When the test is succesful there is only one `Updating Not Supported`:

<details> <summary> Succesful log </summary>

```
<-- {
    "tag": "RspInitialize",
    "contents": {
...................................
    }
}
--> {
    "jsonrpc": "2.0",
    "params": {},
    "method": "initialized"
}
<-- {
    "tag": "ReqRegisterCapability",
    "contents": {
.....................................
    }
}
--> {
    "jsonrpc": "2.0",
    "params": {
        "settings": ""
    },
    "method": "workspace/didChangeConfiguration"
}
<-- {
    "tag": "NotLogMessage",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
            "type": 1,
            "message": "haskell-lsp:configuration parse error. NotificationMessage {_jsonrpc = \"2.0\", _method = WorkspaceDidChangeConfiguration, _params = DidChangeConfigurationParams {_settings = String \"\"}} \"Updating Not supported\""
        },
        "method": "window/logMessage"
    }
}
<-- {
    "tag": "NotLogMessage",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
[DEBUG] finish: kick (took 0.00s)
            "type": 4,
            "message": "Finishing build session(exception: AsyncCancelled)"
        },
        "method": "window/logMessage"
    }
}
<-- {
    "tag": "NotLogMessage",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
            "type": 4,
            "message": "Restarting build session (aborting the previous one took 0.00s)"
        },
        "method": "window/logMessage"
    }
}
--> {
    "jsonrpc": "2.0",
    "params": null,
    "method": "shutdown",
    "id": 0
}
<-- {
    "tag": "ReqWorkDoneProgressCreate",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
            "token": "1"
        },
        "method": "window/workDoneProgress/create",
        "id": 1
    }
}
<-- {
    "tag": "NotWorkDoneProgressBegin",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
            "value": {
                "kind": "begin",
                "title": "Processing"
            },
            "token": "1"
        },
        "method": "$/progress"
    }
}
<-- {
    "tag": "NotWorkDoneProgressEnd",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
            "value": {
                "kind": "end"
            },
            "token": "1"
        },
        "method": "$/progress"
    }
}
<-- {
    "tag": "NotLogMessage",
    "contents": {
        "jsonrpc": "2.0",
        "params": {
            "type": 4,
            "message": "finish: kick (took 0.00s)"
        },
        "method": "window/logMessage"
    }
}
<-- {
    "tag": "RspShutdown",
    "contents": {
        "result": null,
        "jsonrpc": "2.0",
        "id": 0
    }
}
--> {
    "jsonrpc": "2.0",
    "params": [],
    "method": "exit"
}
```

</details>

The intent of the test is "send the config change request and parse the three first logging notifications thereafter".
In both cases there are three logging notifications so i dont know why it get stuck until the timeout.

**Review by pepeiborra — APPROVED** [review]
Awesome job, thank you so much @jneira
